### PR TITLE
Adding retry for target delete

### PR DESCRIPTION
**Pull request by shriharirathod**
- Some cmds can be pending at StorD side.
   TGT force does not take into consideration aborting such IO's.

 - Completion of such IO's needs to be done.
   So adding retry at TGT for deleting target.

Signed-off-by: Shrihari Rathod <shrihari.rathod@primaryio.com>

### CODE REVIEW

File: usr/iscsi/conn.c [hunk 1]
```diff
@@ -95,6 +95,7 @@ void conn_close(struct iscsi_connection *conn)
 
 	conn->closed = 1;
 
+	eprintf("[Sumitk] ep close\n");
```
Comment (sumit-k) [inline usr/iscsi/conn.c]: Remove all messages starting from Sumitk from this file.

File: usr/iscsi/iscsi_tcp.c [hunk 2]
```diff
@@ -278,13 +278,13 @@ static void iscsi_tcp_event_handler(int fd, int events, void *data)
 		iscsi_rx_handler(conn);
 
 	if (conn->state == STATE_CLOSE)
-		dprintf("connection closed\n");
+		dprintf("[Sumitk] connection closed\n");
```
Comment (sumit-k) [inline usr/iscsi/iscsi_tcp.c]: Remove Sumitk keyword.

File: usr/iscsi/iscsid.c [hunk 3]
```diff
@@ -1268,6 +1268,7 @@ static int iscsi_scsi_cmd_done(uint64_t nid, int result, struct scsi_cmd *scmd)
 	 * the response with a little extra code or we can check if this
 	 * task got reassinged to another connection.
 	 */
+
```
Comment (sumit-k) [inline usr/iscsi/iscsid.c]: Remove space.

File: usr/target.c [hunk 4]
```diff
@@ -735,8 +735,10 @@ tgtadm_err tgt_device_destroy(int tid, uint64_t lun, int force)
 		return TGTADM_NO_LUN;
 	}
 
-	if (!list_empty(&lu->cmd_queue.queue) || lu->cmd_queue.active_cmd)
```
Comment (sumit-k) [inline usr/target.c]: Remove messages which we had added.

File: usr/tgtd.c [hunk 5]
```diff
@@ -894,6 +894,8 @@ static int target_delete(const _ha_request *reqp,
 	int rc  = 0;
 	int len = 0;
 	int force = -1;
+	int retry = 12;
```
Comment (sumit-k) [inline usr/tgtd.c]: #define for RETRY values.

File: usr/tgtd.c [hunk 6]
```diff
@@ -910,6 +912,58 @@ static int target_delete(const _ha_request *reqp,
 		return HA_CALLBACK_CONTINUE;
 	}
 
+	/*Unbind before delete*/
+	len = snprintf(cmd, sizeof(cmd),
+			"tgtadm --lld iscsi --mode target --op unbind --tid=%s"
+			" -I ALL", tid);
+	if (len >= sizeof(cmd)) {
+		set_err_msg(resp, TGT_ERR_TOO_LONG,
+			"tgt cmd unbind too long");
```
Comment (sumit-k) [inline usr/tgtd.c]: Return proper error code, what is too long error ?

File: usr/tgtd.c [hunk 7]
```diff
@@ -894,6 +894,8 @@ static int target_delete(const _ha_request *reqp,
 	int rc  = 0;
 	int len = 0;
 	int force = -1;
```
Comment (sumit-k) [inline usr/tgtd.c]: Why initial value is -1 and not simply 1. If possible, make it a bool variable.

File: usr/tgtd.c [hunk 8]
```diff
@@ -925,7 +979,19 @@ static int target_delete(const _ha_request *reqp,
 			"tgt cmd too long");
 		return HA_CALLBACK_CONTINUE;
 	}
```
Comment (sumit-k) [inline usr/tgtd.c]: Where is the else part ? In else there  should be a command without "force" option.

File: usr/tgtd.c [hunk 9]
```diff
@@ -925,7 +979,19 @@ static int target_delete(const _ha_request *reqp,
...
+retry_loop:
 	rc = exec(cmd);
+	if (rc == TGTADM_LUN_ACTIVE ||
+		rc == TGTADM_TARGET_ACTIVE ||
+		rc == TGTADM_DRIVER_ACTIVE ||
+		rc == TGTADM_UNSUPPORTED_OPERATION) {
+		fprintf(stderr, "Retrying for errno: %d\n", rc);
+		sleep(delay);
```
Comment (sumit-k) [inline usr/tgtd.c]: Reduce the delay to 5 secs and correspondingly set the retry value. 10 secs seems a larger wait time for IOs to complete.

File: usr/tgtd.c [hunk 10]
```diff
@@ -925,7 +979,19 @@ static int target_delete(const _ha_request *reqp,
...
+		retry--;
+		if (retry) {
+			goto retry_loop;
```
Comment (sumit-k) [inline usr/tgtd.c]: Use a while loop for retry instead of using goto, which is generally suggested to avoid.

File: usr/bs_hyc.c [hunk 11]
```diff
@@ -136,16 +152,19 @@ static int bs_hyc_cmd_submit(struct scsi_cmd *cmdp)
...
 		assert(0);
 	}
 
 	/* If we got reqid, set it in hyc_cmd */
 	if (hyc_unlikely(reqid == kInvalidRequestID)) {
-		eprintf("request submission got err invalid request\n");
-		target_cmd_io_done(cmdp, SAM_STAT_CHECK_CONDITION);
+		eprintf("request submission got error invalid request" 
+			" size: %lu offset : %"PRIu64" opcode :%u\n", 
+			length, offset, (unsigned int) cmdp->scb[0]);
```
Comment (sumit-k) [inline usr/bs_hyc.c]: We need to test this scenario.. Set the Reqid invalid for write IOs (don't do it for read IOs because that will fail discovery) and see wherther application (may be "dd" command ) is getting completion(error) or not.

File: usr/bs_hyc.c [hunk 12]
```diff
@@ -118,13 +115,32 @@ static int bs_hyc_cmd_submit(struct scsi_cmd *cmdp)
...
+	if (hyc_unlikely(op == WRITE_SAME_OP)) {
+		return -1;
+	} else if (hyc_unlikely(op == UNKNOWN)) {
+		return 0;
 	}
 
 	offset = scsi_cmd_offset(cmdp);
 	length = scsi_cmd_length(cmdp);
 
+	/*
```
Comment (sumit-k) [inline usr/bs_hyc.c]: Need to test this case of zero size IO. Check how to issue barrier IO (check sg_sync scsi command) which are genrally of zero size. We can also try to test using dd by giving it a zero size file for reading (simply touch the file and then try to write it) 

touch /tmp/file1
dd if=/tmp/file1 of=/dev/<discoved_device> count=1 bs=4096 oflag=direct 

If dd is not working then we have to write our own client which opens the device and issue a zero length IO.

File: usr/iscsi/conn.c (hunk 1 above)
Comment (shriharirathod) [reply usr/iscsi/conn.c]: Done

File: usr/iscsi/iscsi_tcp.c (hunk 2 above)
Comment (shriharirathod) [reply usr/iscsi/iscsi_tcp.c]: Done

File: usr/iscsi/iscsid.c (hunk 3 above)
Comment (shriharirathod) [reply usr/iscsi/iscsid.c]: Done

File: usr/target.c (hunk 4 above)
Comment (shriharirathod) [reply usr/target.c]: Done all messages sorted and removed

File: usr/tgtd.c (hunk 5 above)
Comment (shriharirathod) [reply usr/tgtd.c]: Done

File: usr/tgtd.c (hunk 6 above)
Comment (shriharirathod) [reply usr/tgtd.c]: Updated

File: usr/tgtd.c (hunk 8 above)
Comment (shriharirathod) [reply usr/tgtd.c]: Already present Sir.

File: usr/tgtd.c (hunk 10 above)
Comment (shriharirathod) [reply usr/tgtd.c]: Done

File: usr/tgtd.c (hunk 9 above)
Comment (shriharirathod) [reply usr/tgtd.c]: Done

File: usr/tgtd.c (hunk 7 above)
Comment (shriharirathod) [reply usr/tgtd.c]: Done
